Add GitHub Actions to Template

MaxUlysse opened the pull request:
Many thanks to contributing to nf-core/tools!

Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

## PR checklist
 - [ ] This comment contains a description of changes (with reason)
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [ ] `README.md` is updated

**Learn more about contributing:** https://github.com/nf-core/tools/tree/master/.github/CONTRIBUTING.md

MaxUlysse commented:
This PR is based on https://github.com/nf-core/sarek/pull/27

ewels commented:
SVG images are missing... But could probably just use shields.io anyway?

Travis stuff that I think we're still missing:

* Correct branch logic
  * If not from this repo, then must not be against `master`
  * If from repo and against `master`, then coming from `dev`
* That the changelog has been updated (maybe this is only the main tools repo, not pipelines?)

Would be good to have the branch stuff in it's own file / test I think? Would be super quick then and I like having the tests split up like that to make it super clear..

Great work!

ewels commented:
From [here](https://github.com/CultureHQ/github-actions-badge):

> You should put the following into your README which is supported directly by GitHub:
> ```markdown
> [![Actions Status](https://github.com/{owner}/{repo}/workflows/{workflow_name}/badge.svg)](https://github.com/{owner}/{repo}/actions)
> ```

MaxUlysse commented:
> Would be good to have the branch stuff in it's own file / test I think? Would be super quick then and I like having the tests split up like that to make it super clear..

Good point, I totally forgot about that one, I'll add that to the test on my sarek PR

> From [here](https://github.com/CultureHQ/github-actions-badge):

That's exactly the link I used, I just figured one need to replace {owner}, {repo} and {workflow_name}, but I'm guessing it might have been working with the old Actions, and not this one, I'll keep looking

MaxUlysse commented:
Nothing on shields.io yet, if I understood this issue: https://github.com/badges/shields/issues/2574

MaxUlysse commented:
I fixed the badges, and added a test to protect master: https://github.com/nf-core/tools/blob/2a959e24b530c551d52ce42c4c2b22a73e65a55e/nf_core/pipeline-template/%7B%7Bcookiecutter.name_noslash%7D%7D/.github/workflows/branch.yml

It's only triggered for master branch, and check that the GITHUB_AUTHOR is nf-core and that the head branch is dev, which should ensure the same test we were doing.

It should need further testing to verify that, so do not merge until we're sure with the Sarek PR.

MaxUlysse commented:
I would say yes in the long term, but not now, it's still in beta :-D

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/workflows/ci.yml:
```diff
@@ -0,0 +1,20 @@
+      - uses: actions/checkout@v1
+      - name: Install Nextflow
+        run: |
+          export NXF_VER=${{ matrix.nxf_ver }}
```
This is causing the tests to fail, because the double brackets look like a jinja variable.
```suggestion
          {% raw %}export NXF_VER=${{ matrix.nxf_ver }}{% endraw %}
```

apeltzer replied:
Should we merge this into a target branch e.g. `nf-core/tools:github-actions` so that we can add stuff missing? I have some additional checks for PRs coming from `patch` branches which are missing here too but can't commit to Maximes branch 👍

MaxUlysse commented:
Quick question about the branch protection.
So basically we allow PR to master only from dev, except if it's a patch/fix.
Should we have a specific name for such a branch?
Should have some docs about how to patch a pipeline?

ewels commented:
> Should we have a specific name for such a branch?

We do, it's `patch` right?

MaxUlysse commented:
> > Should we have a specific name for such a branch?
> 
> We do, it's patch right?

I know, the important part of the question was:
> Should have some docs about how to patch a pipeline?

apeltzer commented:
Yeah, having some docs on what / how we `patch` a pipeline would be good 👍  

It shouldn't be used too often, though I hope 😆

MaxUlysse commented:
I updated the `CONTRIBUTING.md` and related files.

MaxUlysse commented:
Anyone has any idea why tests are failing on that one?

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/CONTRIBUTING.md:
```diff
@@ -1,47 +1,53 @@
+However, don't be put off by this template - other more general issues and suggestions are welcome!
+Contributions to the code are even more welcome ;)
 
+> If you need help using or modifying {{ cookiecutter.name }} then the best place to ask is on the nf-core Slack [#{{ cookiecutter.cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).
```
```suggestion
> If you need help using or modifying {{ cookiecutter.name }} then the best place to ask is on the nf-core Slack [#{{ cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/CONTRIBUTING.md:
```diff
@@ -1,47 +1,53 @@
 ## Getting help
-For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://github.com/{{ cookiecutter.name }}#documentation) and don't hesitate to get in touch on the [{{ cookiecutter.name }} pipeline channel](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) on [Slack](https://nf-co.re/join/slack).
+
+For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://github.com/{{ cookiecutter.name }}#documentation) and don't hesitate to get in touch on the nf-core Slack [#{{ cookiecutter.cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).
```
```suggestion
For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://github.com/{{ cookiecutter.name }}#documentation) and don't hesitate to get in touch on the nf-core Slack [#{{ cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).
```

MaxUlysse commented:
:man_facepalming:
Thanks a lot, I'm sure I was about to get there...

MaxUlysse commented:
@ewels what do you think about the mergability of this one?

ewels commented:
I'll review this if you guys review https://github.com/nf-core/tools/pull/467 and https://github.com/nf-core/tools/pull/465 😉

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/CONTRIBUTING.md:
```diff
@@ -1,45 +1,53 @@
+
+## Patch
+
+When patching a release, please work on your fork on a new branch named `patch`
```
I think we maybe need a little more here explaining what a patch is, when you would patch a release and make it abundantly clear that it should never happen 😉

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/CONTRIBUTING.md:
```diff
@@ -1,45 +1,53 @@
 ## Getting help
-For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://github.com/{{ cookiecutter.name }}#documentation) and don't hesitate to get in touch on the [{{ cookiecutter.name }} pipeline channel](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) on [Slack](https://nf-co.re/join/slack).
+
+For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://github.com/{{ cookiecutter.name }}#documentation) and don't hesitate to get in touch on the nf-core Slack [#{{ cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).
```
```suggestion
For further information/help, please consult the [{{ cookiecutter.name }} documentation](https://nf-co.re/{{ cookiecutter.name }}/docs) and don't hesitate to get in touch on the nf-core Slack [#{{ cookiecutter.short_name }}](https://nfcore.slack.com/channels/{{ cookiecutter.short_name }}) channel ([join our Slack here](https://nf-co.re/join/slack)).
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
- - Version [e.g. 7, 10.13.6, 18.3...]
+## System
+
+- Hardware: [e.g. HPC, Desktop, Cloud...]
```
```suggestion
- Hardware: <!-- [e.g. HPC, Desktop, Cloud...] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
+## System
+
+- Hardware: [e.g. HPC, Desktop, Cloud...]
+- Executor: [e.g. slurm, local, awsbatch...]
```
```suggestion
- Executor: <!-- [e.g. slurm, local, awsbatch...] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
+
+- Hardware: [e.g. HPC, Desktop, Cloud...]
+- Executor: [e.g. slurm, local, awsbatch...]
+- OS: [e.g. CentOS Linux, macOS, Linux Mint...]
```
```suggestion
- OS: <!-- [e.g. CentOS Linux, macOS, Linux Mint...] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
+- Hardware: [e.g. HPC, Desktop, Cloud...]
+- Executor: [e.g. slurm, local, awsbatch...]
+- OS: [e.g. CentOS Linux, macOS, Linux Mint...]
+- Version [e.g. 7, 10.13.6, 18.3...]
```
```suggestion
- Version <!-- [e.g. 7, 10.13.6, 18.3...] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
+
+## Nextflow Installation
+
+- Version: [e.g. 0.31.0]
```
```suggestion
- Version: <!-- [e.g. 19.10.0] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
 
-#### Nextflow Installation:
- - Version: [e.g. 0.31.0]
+- Engine: [e.g. Conda, Docker or Singularity]
```
```suggestion
- Engine: <!-- [e.g. Conda, Docker or Singularity] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
-#### Nextflow Installation:
- - Version: [e.g. 0.31.0]
+- Engine: [e.g. Conda, Docker or Singularity]
+- version: [e.g. 1.0.0]
```
```suggestion
- version: <!-- [e.g. 1.0.0] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,31 +1,42 @@
- - Version: [e.g. 0.31.0]
+- Engine: [e.g. Conda, Docker or Singularity]
+- version: [e.g. 1.0.0]
+- Image tag: [e.g. {{ cookiecutter.name_docker }}:1.0.0]
```
```suggestion
- Image tag: <!-- [e.g. {{ cookiecutter.name_docker }}:1.0.0] -->
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/workflows/branch.yml:
```diff
@@ -0,0 +1,15 @@
+  test:
+    runs-on: ubuntu-18.04
+    steps:
+      # PRs are only ok if coming from an nf-core dev branch
```
```suggestion
      # PRs are only ok if coming from an nf-core `dev` branch or a fork `patch` branch
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/workflows/branch.yml:
```diff
@@ -0,0 +1,15 @@
+name: nf-core branch protection
+# This workflow is triggered on PRs to master branch on the repository
```
```suggestion
# This workflow is triggered on PRs to master branch on the repository
# It fails when someone tries to make a PR against the nf-core `master` branch instead of `dev`
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/workflows/ci.yml:
```diff
@@ -0,0 +1,20 @@
+name: nf-core CI
+# This workflow is triggered on pushes and PRs to the repository.
```
```suggestion
# This workflow is triggered on pushes and PRs to the repository.
# It runs the pipeline with the minimal test dataset to check that it completes without any syntax errors
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/workflows/ci.yml:
```diff
@@ -0,0 +1,20 @@
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        nxf_ver: ['19.10.0', '']
```
```suggestion
        # Nextflow versions: check pipeline minimum and current latest
        nxf_ver: ['19.10.0', '']
```

ewels commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/workflows/linting.yml:
```diff
@@ -0,0 +1,40 @@
+name: nf-core linting
+# This workflow is triggered on pushes and PRs to the repository.
```
```suggestion
# This workflow is triggered on pushes and PRs to the repository.
# It runs the `nf-core lint` and markdown lint tests to ensure that the code meets the nf-core guidelines
```

MaxUlysse commented on nf_core/pipeline-template/{{cookiecutter.name_noslash}}/.github/CONTRIBUTING.md:
```diff
@@ -1,45 +1,53 @@
+
+## Patch
+
+When patching a release, please work on your fork on a new branch named `patch`
```
```suggestion
: warning: Only in the unlikely and regretful event of a release happening with a bug.

* On your own fork, make a new branch `patch` based on `upstream/master`.
* Fix the bug, and bump version (X.Y.Z+1).
* A PR should be made on `master` from patch to directly this particular bug.
```

MaxUlysse replied:
@ewels What do you think of that?

ewels replied:
Looks good!
